test/cli/test/bail-race.test.ts

```diff
@@ -0,0 +1,44 @@
+import { resolve } from 'pathe'
+import { expect, test } from 'vitest'
+import { createVitest } from 'vitest/node'
+
+test('cancels previous run before starting new one', async () => {
+  const results: string[] = []
+
+  const vitest = await createVitest('test', {
+    root: resolve(import.meta.dirname, '../fixtures/bail-race'),
+    bail: 1,
+    maxWorkers: 1,
+    watch: false,
+    reporters: [{
+      onTestCaseResult(testCase) {
+        const result = testCase.result()
+
+        results.push(`${result.state}${result.errors ? `: ${result.errors?.[0].message}` : ''}`)
+      },
+    }],
+  })
+
+  let rounds = 0
+
+  while (vitest.state.errorsSet.size === 0) {
+    await vitest.start()
+
+    if (rounds >= 2) {
+      break
+    }
+
+    rounds++
+  }
+
+  expect(results).toMatchInlineSnapshot(`
+    [
```

Comment:
This structure is a bit confusing to me, it doesn't have test names. `test-utils` has a `buildTree` function somewhere, maybe use it here instead? I think at the moment it is only returned from `runInlinedTests`, but you can add a return in `runVitest`